src/navigation/gate.py

```diff
@@ -70,11 +70,60 @@ def spider_gate_trajectory(cls, approach_distance: float, gate: Gate, rover_posi
         closest_approach_point = possible_approach_points[approach_idx]
         victory_point = possible_approach_points[1 - approach_idx]
 
+        coordinates = GateTrajectory.gateSelectPath(
+            rover_position, closest_prep_point, closest_approach_point, center, victory_point, gate
+        )
+
         # put the list of coordinates together
-        coordinates = np.array([closest_prep_point, closest_approach_point, victory_point])
-        coordinates = np.hstack((coordinates, np.zeros(coordinates.shape[0]).reshape(-1, 1)))
         return GateTrajectory(coordinates)
 
+    def gateSelectPath(
+        rover_position: np.ndarray, pt1: np.ndarray, pt2: np.ndarray, pt3: np.ndarray, pt4: np.ndarray, gate: Gate
+    ):
+        # Get the shapes of both the posts
+        postOneShape, postTwoShape = gate.getPostGeoShape()
+
+        # Get points for path
+        pt0 = np.array([rover_position[0], rover_position[1]])
+
+        # Get paths
+        (
+            pathOne,
+            pathTwo,
+        ) = GateTrajectory.pathLineString(pt0, pt2, pt3, pt4)
+
+        # Check if the lines intersect
+        # First check if the path1 hits either posts
+        if GateTrajectory.lineIntersectCheck(pathOne, postOneShape) or GateTrajectory.lineIntersectCheck(
+            pathOne, postTwoShape
+        ):
+            if GateTrajectory.lineIntersectCheck(pathTwo, postOneShape) or GateTrajectory.lineIntersectCheck(
```

Comment:
I think the code could be structured a bit cleaner here. We could do something like

```
all_pts = [prep, approach, center, victory]
start_index = 2
path = make_path_obj(all_pts[start_index:])
while path.intersects(postOneShape) or path.intersects(postTwoShape):
     start_index -= 1
     if start_index == 0:
          break
     path = make_path_obj(all_pts[start_index:])
coordinates = np.array(path[stard_index : ])
coordinates = np.hstack(coordinates, np.zeros(coordinates.shape[0].reshape(-1, 1)))
```

You'd need to define the function make_path_obj but all it would do is actually create the path object (LineString) from the rovers position to the first point in the list and then the next point and the next (you can do this with a for loop too)

The way it works is it starts with just the center and the victory point and as long as one of the posts is intersecting with the path, it moves the start index back by one and reconstructs the path object with one more segment on the path. If the start index goes to zero, then we can't optimize it anymore since that's just the original path.

Reply:
what this also does is it provides flexibility in the length of our path and also makes it relatively easy to change the condition that we're actually optimizing on